caproto/ioc_examples/lakeshore.py

```diff
@@ -0,0 +1,390 @@
+        return self._output
+
+    @property
+    def feedback(self):
+        return self._feedback
+
+    @property
+    def setpoint(self):
+        return self._setpoint
+
+    @setpoint.setter
+    def setpoint(self, value):
+        """
+        Always set ramping to true at the same time as
+        the setpoint change. Is will avoid a race condition,
+        and will allow the client to just check for ramping = False
+        to determine completion.
+        """
+        self._setpoint_target = value
+        self._ramping = True
+        if self.ramp_rate is None:
+            self._setpoint = value
+
+    @property
+    def ramping(self):
+        return self._ramping
+
+    def stop(self):
+        self._run = False
+
+    def _executor(self):
+        iteration_time = 0.1
+        while self._run:
+            self._feedback = self._get_feedback()
+            self._output = self.__call__(self._get_feedback())
+            self.history["output"].append(self._output)
+            self.history["feedback"].append(self._feedback)
+            self.history["setpoint"].append(self.setpoint)
+            self.history["timestamp"].append(time.time())
+            self._set_output(self._output)
+
+            # Ramping logic.
+            remaining = self._setpoint_target - self.setpoint
+            self._ramping = bool(remaining)
+            if isinstance(self.ramp_rate, (int, float)):
+                if remaining > 0:
+                    self._setpoint += min(
+                        self.ramp_rate * iteration_time, abs(remaining)
+                    )
+                elif remaining < 0:
+                    self._setpoint -= min(
+                        self.ramp_rate * iteration_time, abs(remaining)
+                    )
+            elif self.ramp_rate is None and remaining != 0:
+                self._setpoint = self._setpoint_target
+
+            time.sleep(iteration_time)
+
+
+internal_process = contextvars.ContextVar("internal_process", default=False)
+
+
+def no_reentry(func):
+    """
+    This is needed for put completion.
+    """
+
+    @functools.wraps(func)
+    async def inner(*args, **kwargs):
+        if internal_process.get():
+            return
+        try:
+            internal_process.set(True)
+            return await func(*args, **kwargs)
+        finally:
+            internal_process.set(False)
+
+    return inner
+
+
+class LakeshoreIOC(PVGroup):
+    """
+    Simulated Lakeshore IOC with put completion on the setpoint.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._sample = ThermalMaterial()
+
+        self._temperature_controller = PIDController(
+            lambda: self._sample.temperature,
+            self._sample.set_heater_power,
+            Kp=1,
+            Ki=0.1,
+            Kd=0.05,
+            setpoint=150,
+        )
+
+    Kp = pvproperty(value=0, dtype=float, name="Kp", doc="PID parameter Kp")
+
+    @Kp.getter
```

Comment:
Oh okay, that makes sense. 

I followed this example: https://github.com/caproto/caproto/blob/master/caproto/ioc_examples/too_clever/trigger_with_pc.py#L107

Do you have any suggestions about how I can set update Kp? Do I need another thread to read from self._temperature_controler.Kp and write to self.Kp?

Reply:
Those examples need either reworking or removal - I'll make an issue for that, thanks.

I'd recommend a few things:
* Removing threading entirely -> use asyncio for updates
* Keeping the canonical source of parameter values inside the IOC and moving the simulation logic to a single spot. * Take in new parameters and update all simulated values in a single step such as:

```python

    async def simulate(self, dt: float):
        # Update internal parameters
        self._temperature_controller.Kd = self.Kd.value
        ...
        # Simulate
        self._temperature_controller.simulate(dt)
        ...
        # Update outputs
        await self.feedback.write(self._temperature_controller.feedback)
        await self.output.write(self._temperature_controller.output)
```

I had a few minutes this morning to put the words into code - https://github.com/caproto/caproto/commit/b0362a9fcecf88171ae0422795fe7e2c70b918d5
 No guarantees that it will actually work but I think the concept should be relatively clear, at least.
Feel free to do with it what you will.

Reply:
I avoided putting simulation logic in the IOC, because I wanted it to model a situation where you have a physical temperature controller that is totally separate from the IOC.  And I wanted to understand what the IOC would look like if it had to communicate with a physical controller.  So someone could walk up to the controller and change the gains directly on the controller, or change the gains on the controller through some other method, like CSS.  If this IOC were to work for a physical temperature controller, I guess it would need to poll it?

Reply:
I don't think there's a large difference between the following from a user standpoint.
1. A thread that loops and runs `simulate(dt)`
2. Let the user call `simulate(dt)` in their own event loop

However, one makes for a clean caproto IOC and one does not.

If this example had to communicate with a real controller, then the source of your information would change. No longer would caproto be the primary source of Kp/Ki/Kd setpoint information, but the controller would.

Typically, in normal EPICS IOCs, you would have a polling loop that reads the current state of a controller into a set of records. You would also have setpoint records that cause a write to the controller on-demand (i.e., when the user does a `caput`).

For caproto, you'd need a polling loop in a separate task. And the split-up of setpoint and readback values would look something like this:
```python
    Kp = pvproperty(value=1.0, doc="PID parameter Kp")
    Kp_rbv = pvproperty(value=1.0, doc="PID parameter Kp (readback value)")
    Ki = pvproperty(value=0.1, doc="PID parameter Ki")
    Kp_rbv = pvproperty(value=1.0, doc="PID parameter Kp (readback value)")
    Kd = pvproperty(value=0, doc="PID parameter Kd")
    Kd_rbv = pvproperty(value=1.0, doc="PID parameter Kp (readback value)")
```
The user writes to `Kp` to change the setpoint on the controller.
The controller current value is read back to `Kp_rbv`.

This often confuses people new to EPICS - "why do I need two PVs for a single value?" Well, because your representation of the setpoint and the controller's representation of the setpoint can and will differ (at startup, if the controller reboots, in the time period between the user writing to setpoint and the controller accepting it, if the controller rejects a certain setpoint value, if there is a lack of communication between the IOC and controller, etc.)

Reply:
What if caproto has a decorator like this?

```
@Kp_rbv.poller(5) # time in seconds
async def Kp_rbv(self, instance):
    return await device.read_Kp()
```

And the IOC could call all of the pollers at the defined interval.

Reply:
You can do just that:
https://caproto.github.io/caproto/v1.0.0/iocs.html#scan-hook

In the scan hook, `await self.Kp_rbv.write(await device.read_Kp())` - which could be fine. But you may have to take care of locking the device on your own (what happens if your independent polling methods for Kp and Ki and Kd all happen at the same time and jumble the responses?), whereas with a single poll method you combine it into step-by-step polling of each item (read kp, read ki, and then read kd, repeat).

Reply:
Thanks for the advice! Ill make some updates

Reply:
Does it look good now?